### PR TITLE
[datasetexporter]: Upgrade library to 0.17.0

### PR DESCRIPTION
**Pull request by martin-majlis-s1**
**Description:** Upgrade to new version of the library.

This PR is implementing following issues:

* #27650 - metrics are not collected via open telemetry, so they can be monitored. It's better version of the previous PR #27487 which was not working. 
* #27652 - it's configurable with the `debug` option whether `session_key` is included or not

Other change is that fields that are specified as part of the `group_by` configuration are now transferred as part of the session info.

**Link to tracking Issue:** #27650, #27652

**Testing:** 

1. Build docker image - make docker-otelcontribcol
2. Checkout https://github.com/open-telemetry/opentelemetry-demo
3. Update configuration in `docker-compose.yaml` and in the  `src/otelcollector/otelcol-config.yml`:
  * In `docker-compose.yaml` switch image to the newly build one in step 1
  * In `docker-compose.yaml` enable feature gate for collecting metrics - `--feature-gates=telemetry.useOtelForInternalMetrics`
  * In `src/otelcollector/otelcol-config.yml` enable metrics scraping by prometheus
  * In `src/otelcollector/otelcol-config.yml` add configuration for dataset
```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 001f7c8..d7edd0d 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -646,14 +646,16 @@ services:

   # OpenTelemetry Collector
   otelcol:
-    image: otel/opentelemetry-collector-contrib:0.86.0
+    image: otelcontribcol:latest
     container_name: otel-col
     deploy:
       resources:
         limits:
           memory: 125M
     restart: unless-stopped
-    command: [ "--config=/etc/otelcol-config.yml", "--config=/etc/otelcol-config-extras.yml" ]
+    command: [ "--config=/etc/otelcol-config.yml", "--config=/etc/otelcol-config-extras.yml", "--feature-gates=telemetry.useOtelForInternalMetrics" ]
     volumes:
       - ./src/otelcollector/otelcol-config.yml:/etc/otelcol-config.yml
       - ./src/otelcollector/otelcol-config-extras.yml:/etc/otelcol-config-extras.yml
diff --git a/src/otelcollector/otelcol-config.yml b/src/otelcollector/otelcol-config.yml
index f2568ae..9944562 100644
--- a/src/otelcollector/otelcol-config.yml
+++ b/src/otelcollector/otelcol-config.yml
@@ -15,6 +15,14 @@ receivers:
     targets:
       - endpoint: http://frontendproxy:${env:ENVOY_PORT}

+  prometheus:
+    config:
+      scrape_configs:
+        - job_name: 'otel-collector'
+          scrape_interval: 5s
+          static_configs:
+            - targets: ['0.0.0.0:8888']
+
 exporters:
   debug:
   otlp:
@@ -29,6 +37,22 @@ exporters:
     endpoint: "http://prometheus:9090/api/v1/otlp"
     tls:
       insecure: true
+  logging:
+  dataset:
+    api_key: API_KEY
+    dataset_url: https://SERVER.scalyr.com
+    debug: true
+    buffer:
+      group_by:
+        - resource_name
+        - resource_type
+    logs:
+      export_resource_info_on_event: true
+    server_host:
+      server_host: Martin
+      use_hostname: false
+  dataset/aaa:
+    api_key: API_KEY
+    dataset_url: https://SERVER.scalyr.com
+    debug: true
+    buffer:
+      group_by:
+        - resource_name
+        - resource_type
+    logs:
+      export_resource_info_on_event: true
+    server_host:
+      server_host: MartinAAA
+      use_hostname: false

 processors:
   batch:
@@ -47,6 +71,11 @@ processors:
           - set(description, "") where name == "queueSize"
           # FIXME: remove when this issue is resolved: https://github.com/open-telemetry/opentelemetry-python-contrib/issues/1958
           - set(description, "") where name == "http.client.duration"
+  attributes:
+    actions:
+      - key: otel.demo
+        value: 29446
+        action: upsert

 connectors:
   spanmetrics:
@@ -55,13 +84,13 @@ service:
   pipelines:
     traces:
       receivers: [otlp]
-      processors: [batch]
-      exporters: [otlp, debug, spanmetrics]
+      processors: [batch, attributes]
+      exporters: [otlp, debug, spanmetrics, dataset, dataset/aaa]
     metrics:
-      receivers: [httpcheck/frontendproxy, otlp, spanmetrics]
+      receivers: [httpcheck/frontendproxy, otlp, spanmetrics, prometheus]
       processors: [filter/ottl, transform, batch]
       exporters: [otlphttp/prometheus, debug]
     logs:
       receivers: [otlp]
-      processors: [batch]
-      exporters: [otlp/logs, debug]
+      processors: [batch, attributes]
+      exporters: [otlp/logs, debug, dataset, dataset/aaa]
```
4. Run the demo - `docker compose up --abort-on-container-exit`
5. Check, that metrics are in Grafana - http://localhost:8080/grafana/explore? 
Screenshot 2023-11-27 at 12 29 29
6. Check some metrics
Screenshot 2023-11-22 at 14 06 56
Screenshot 2023-11-27 at 12 59 10
7. Check that data are available in dataset Screenshot 2023-11-22 at 13 33 50 

**Documentation:** 

**Library changes:**
* Group By & Debug - https://github.com/scalyr/dataset-go/pull/62
* Metrics  - https://github.com/scalyr/dataset-go/pull/61

### CODE REVIEW

**Review by atoulme — APPROVED** [review]

File: exporter/datasetexporter/factory_test.go [hunk 1]
```diff
@@ -160,7 +161,7 @@ func createExporterTests() []CreateTest {
...
+			expectedError: fmt.Errorf("cannot get DataSetExpoter: cannot convert config: DatasetURL: ; APIKey:  (0); Debug: false; BufferSettings: {MaxLifetime:0s GroupBy:[] RetryInitialInterval:0s RetryMaxInterval:0s RetryMaxElapsedTime:0s RetryShutdownTimeout:0s}; LogsSettings: {ExportResourceInfo:false ExportResourcePrefix: ExportScopeInfo:false ExportScopePrefix: DecomposeComplexMessageField:false DecomposedComplexMessagePrefix: exportSettings:{ExportSeparator: ExportDistinguishingSuffix:}}; TracesSettings: {exportSettings:{ExportSeparator: ExportDistinguishingSuffix:}}; ServerHostSettings: {UseHostName:false ServerHost:}; RetrySettings: {Enabled:false InitialInterval:0s RandomizationFactor:0 Multiplier:0 MaxInterval:0s MaxElapsedTime:0s}; QueueSettings: {Enabled:false NumConsumers:0 QueueSize:0 StorageID:<nil>}; TimeoutSettings: {Timeout:0s}; config is not valid: api_key is required"),
```
[the remark this answers is not in the document]
Comment (martin-majlis-s1) [reply exporter/datasetexporter/factory_test.go:164]: Wow!!! Really great catch! It's there from the beginning. :) I will fix this. :)

File: exporter/datasetexporter/README.md [hunk 2]
```diff
@@ -280,3 +279,41 @@ Based on the given configuration and scenarios, here's the expected behavior:
 5. Resource: `{}`, Log: `{'attribute.foo': 'Bar'}`, Env: `SERVER_HOST=''`, Hostname: `ip-172-31-27-19`
    * Since the attribute `container_id` is not set and the environmental variable `SERVER_HOST` is empty, the `hostname` of the node (`ip-172-31-27-19`) will be used as the fallback value for `serverHost`.
    * Used `serverHost` will be `ip-172-31-27-19`.
+
+## Metrics
+
+To enable metrics you have to:
+1. Run collector with enabled feature gate `telemetry.useOtelForInternalMetrics`. This can be done by executing it with one additional parameter - `--feature-gates=telemetry.useOtelForInternalMetrics`.
+2. Enable metrics scraping as part of the configuration with and add receiver into services:
```
Comment (astencel-sumo) [inline exporter/datasetexporter/README.md:287]:
```suggestion
2. Enable metrics scraping as part of the configuration and add receiver into services:
```

File: .chloggen/datasetexporter-update-to-0.16.0.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: datasetexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Collect usage metrics with Otel and sent grouped attributes in session info.
```
Comment (astencel-sumo) [inline .chloggen/datasetexporter-update-to-0.16.0.yaml:10]:
```suggestion
note: Collect usage metrics with Otel and send grouped attributes in session info.
```

File: exporter/datasetexporter/README.md [hunk 4]
```diff
@@ -280,3 +279,41 @@ Based on the given configuration and scenarios, here's the expected behavior:
...
+   service:
+     pipelines:
+       metrics:
+         # add prometheus among metrics receivers
+         receivers: [prometheus]
+         processors: [batch]
+         exporters: [otlphttp/prometheus, debug]
+   ```
+
+#### Available Metrics
```
Comment (astencel-sumo) [inline exporter/datasetexporter/README.md:307]:
```suggestion
### Available Metrics
```

**Comment by astencel-sumo** [comment]
I'm worried that these metrics don't look like other metrics exposed by other components of the collector:

```console
$ curl localhost:8888/metrics
# HELP otelcol_dataset_eventsEnqueued 
# TYPE otelcol_dataset_eventsEnqueued gauge
otelcol_dataset_eventsEnqueued 31
# HELP otelcol_dataset_eventsProcessed 
# TYPE otelcol_dataset_eventsProcessed gauge
otelcol_dataset_eventsProcessed 31
(...)
```

1. The metric names are partially `snake_case` and partially `camelCase`, while other Otelcol metrics only use `snake_case`.
2. The metrics don't have any attributes. What happens if I have two distinct exporter instances configured?

**Comment by mx-psi** [comment]
```
datasetexporter/datasetexporter.go:9: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
	"github.com/scalyr/dataset-go/pkg/meter_config"
datasetexporter/datasetexporter.go:17: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
	"github.com/scalyr/dataset-go/pkg/client"
datasetexporter/datasetexporter.go:14: File is not `goimports`-ed with -local github.com/open-telemetry/opentelemetry-collector-contrib (goimports)
```

File: exporter/datasetexporter/config.go [hunk 5]
```diff
@@ -182,14 +186,15 @@ func (c *Config) Validate() error {
 func (c *Config) String() string {
 	s := ""
 	s += fmt.Sprintf("%s: %s; ", "DatasetURL", c.DatasetURL)
+	s += fmt.Sprintf("%s: %s (%d); ", "APIKey", strings.Repeat("*", len(c.APIKey)), len(c.APIKey))
```
Comment (codeboten) [inline exporter/datasetexporter/config.go:189]: a configopaque.String will by default be replaced with `[Redacted]`, is there a reason to override the behaviour here?

**Review by codeboten — COMMENTED** [review]
Change looks good, just one question

File: exporter/datasetexporter/config.go (hunk 5 above)
Comment (martin-majlis-s1) [reply exporter/datasetexporter/config.go:189]: I haven't realised, that this is how `configopaque.String` works. Then there is no need to do this replacements. :) 
Thanks :)

Comment (martin-majlis-s1) [reply exporter/datasetexporter/config.go:189]: @codeboten : Thanks. But when I have tried it out - it's returning the real value. I have to use `apiKey, _ := c.APIKey.MarshalText()` - which is not too much better. I will replace it, so it's more consistent with other places.


```diff
--- FAIL: TestConfigString (0.00s)
    config_test.go:143:
                Error Trace:    /Users/martin.majlis/development/scalyr-org/opentelemetry-collector-contrib/exporter/datasetexporter/config_test.go:143
                Error:          Not equal:
                                expected: "DatasetURL: https://example.com; APIKey: ****** (6); Debug: true; BufferSettings: {MaxLifetime:123ns GroupBy:[field1 field2] RetryInitialInterval:0s RetryMaxInterval:0s RetryMaxElapsedTime:0s RetryShutdownTimeout:0s}; LogsSettings: {ExportResourceInfo:true ExportResourcePrefix:AAA ExportScopeInfo:true ExportScopePrefix:BBB DecomposeComplexMessageField:true DecomposedComplexMessagePrefix:EEE exportSettings:{ExportSeparator:CCC ExportDistinguishingSuffix:DDD}}; TracesSettings: {exportSettings:{ExportSeparator:TTT ExportDistinguishingSuffix:UUU}}; ServerHostSettings: {UseHostName:false ServerHost:foo-bar}; RetrySettings: {Enabled:true InitialInterval:5s RandomizationFactor:0.5 Multiplier:1.5 MaxInterval:30s MaxElapsedTime:5m0s}; QueueSettings: {Enabled:true NumConsumers:10 QueueSize:1000 StorageID:<nil>}; TimeoutSettings: {Timeout:5s}"
                                actual  : "DatasetURL: https://example.com; APIKey: secret (6); Debug: true; BufferSettings: {MaxLifetime:123ns GroupBy:[field1 field2] RetryInitialInterval:0s RetryMaxInterval:0s RetryMaxElapsedTime:0s RetryShutdownTimeout:0s}; LogsSettings: {ExportResourceInfo:true ExportResourcePrefix:AAA ExportScopeInfo:true ExportScopePrefix:BBB DecomposeComplexMessageField:true DecomposedComplexMessagePrefix:EEE exportSettings:{ExportSeparator:CCC ExportDistinguishingSuffix:DDD}}; TracesSettings: {exportSettings:{ExportSeparator:TTT ExportDistinguishingSuffix:UUU}}; ServerHostSettings: {UseHostName:false ServerHost:foo-bar}; RetrySettings: {Enabled:true InitialInterval:5s RandomizationFactor:0.5 Multiplier:1.5 MaxInterval:30s MaxElapsedTime:5m0s}; QueueSettings: {Enabled:true NumConsumers:10 QueueSize:1000 StorageID:<nil>}; TimeoutSettings: {Timeout:5s}"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -DatasetURL: https://example.com; APIKey: ****** (6); Debug: true; BufferSettings: {MaxLifetime:123ns GroupBy:[field1 field2] RetryInitialInterval:0s RetryMaxInterval:0s RetryMaxElapsedTime:0s RetryShutdownTimeout:0s}; LogsSettings: {ExportResourceInfo:true ExportResourcePrefix:AAA ExportScopeInfo:true ExportScopePrefix:BBB DecomposeComplexMessageField:true DecomposedComplexMessagePrefix:EEE exportSettings:{ExportSeparator:CCC ExportDistinguishingSuffix:DDD}}; TracesSettings: {exportSettings:{ExportSeparator:TTT ExportDistinguishingSuffix:UUU}}; ServerHostSettings: {UseHostName:false ServerHost:foo-bar}; RetrySettings: {Enabled:true InitialInterval:5s RandomizationFactor:0.5 Multiplier:1.5 MaxInterval:30s MaxElapsedTime:5m0s}; QueueSettings: {Enabled:true NumConsumers:10 QueueSize:1000 StorageID:<nil>}; TimeoutSettings: {Timeout:5s}
                                +DatasetURL: https://example.com; APIKey: secret (6); Debug: true; BufferSettings: {MaxLifetime:123ns GroupBy:[field1 field2] RetryInitialInterval:0s RetryMaxInterval:0s RetryMaxElapsedTime:0s RetryShutdownTimeout:0s}; LogsSettings: {ExportResourceInfo:true ExportResourcePrefix:AAA ExportScopeInfo:true ExportScopePrefix:BBB DecomposeComplexMessageField:true DecomposedComplexMessagePrefix:EEE exportSettings:{ExportSeparator:CCC ExportDistinguishingSuffix:DDD}}; TracesSettings: {exportSettings:{ExportSeparator:TTT ExportDistinguishingSuffix:UUU}}; ServerHostSettings: {UseHostName:false ServerHost:foo-bar}; RetrySettings: {Enabled:true InitialInterval:5s RandomizationFactor:0.5 Multiplier:1.5 MaxInterval:30s MaxElapsedTime:5m0s}; QueueSettings: {Enabled:true NumConsumers:10 QueueSize:1000 StorageID:<nil>}; TimeoutSettings: {Timeout:5s}
                Test:           TestConfigString
```

**Comment by astencel-sumo** [comment]
> @astencel-sumo : I have updated the library - [scalyr/dataset-go#63](https://github.com/scalyr/dataset-go/pull/63).
> 
> I have also updated some screenshots in the description to show the functionality.

This is what I'm now seeing in the metrics:

```console
otelcol_dataset_events_enqueued{entity="logs",name="ds1"} 31
otelcol_dataset_events_enqueued{entity="logs",name="ds2"} 31
otelcol_dataset_events_processed{entity="logs",name="ds1"} 31
otelcol_dataset_events_processed{entity="logs",name="ds2"} 31
```

This is when I have two exporters defined like this:

```yaml
exporters:
  dataset/ds1:
    api_key: abc123456
    dataset_url: http://localhost:1234
  dataset/ds2:
    api_key: def
    dataset_url: http://localhost:5678
service:
  pipelines:
    logs/p1:
      exporters:
        - dataset/ds1
        - dataset/ds2
```

It's not ideal, but the exporter names are there in some form. If you're prepared to possibly make changes to these metrics in the future than I guess this is a step in the right direction.

We should probably also make it easier for components to expose metrics in the right format, and have that documented.

**Review by andrzej-stencel — APPROVED** [review]

File: exporter/datasetexporter/config.go (hunk 5 above)
Comment (codeboten) [reply exporter/datasetexporter/config.go]: > But when I have tried it out - it's returning the real value.
This is surprising to me... maybe @mx-psi can help here. Will not block this PR on this though.

**Review by codeboten — APPROVED** [review]

**Pull request merged**